Goroutines optional

zackwine opened the pull request:
- Add option to set goroutine flushes concurrency level
- Make the default use non-concurrent flushes
- Update progressive backoff logic to handle log spikes

zackwine committed:
Fix retry logic for concurrent flushes.

Fixes duplicate logs when kinesis limits are exceeded.

Reduces the likelihood of dropped logs when kinesis provides backpressure (improvements in retries).

Adds log messages to indicate when and how many records are dropped due to retry timeouts.

Signed-off-by: Zack Wine <zwine@synamedia.com>

PettitWesley committed:
Remove exponential backoff code (#27)

PettitWesley committed:
Bump version to 1.2.2

zackwine committed:
Address code review.  Rename backoff to currentRetries for clarity.

Signed-off-by: Zack Wine <zwine@synamedia.com>

zackwine committed:
Address code review.  Make limitsExceeded var to ensure message is valid.

Signed-off-by: Zack Wine <zwine@synamedia.com>

zackwine committed:
Add concurrency option, and make concurrent goroutine flushes optional.

zackwine committed:
Merge branch 'goroutines' into goroutines-optional

zackwine commented:
@PettitWesley Please review these changes and let me know if this fits what you had in mind.

PettitWesley commented on fluent-bit-kinesis.go:
```diff
@@ -33,8 +33,7 @@ const (
 const (
-	retries              = 6
-	concurrentRetryLimit = 4
+	retries = 6
```
Retries should be configurable instead of hard-coded. I also think 6 is a bit high for that default value...

PettitWesley commented on README.md:
```diff
@@ -16,7 +16,8 @@ If you think you’ve found a potential security issue, please do not post it in
 * `time_key`: Add the timestamp to the record under this key. By default the timestamp from Fluent Bit will not be added to records sent to Kinesis.
-* `time_key_format`: [strftime](http://man7.org/linux/man-pages/man3/strftime.3.html) compliant format string for the timestamp; for example, `%Y-%m-%dT%H:%M:%S%z`. This option is used with `time_key`. 
+* `time_key_format`: [strftime](http://man7.org/linux/man-pages/man3/strftime.3.html) compliant format string for the timestamp; for example, `%Y-%m-%dT%H:%M:%S%z`. This option is used with `time_key`.
+* `concurrency`: Specify a limit of concurrent go routines for flushing records to kinesis.  By default `concurrency` is set to `0` and records are flushed in the context of the calling fluentbit coroutine.  If this value is set to `4` for example then calls to Flush records from fluentbit will spawn concurrent go routines until the limit of `4` concurrent go routines are running.  Once the `conc ...
```
> By default `concurrency` is set to `0` and records are flushed in the context of the calling fluentbit coroutine.

Your average Fluent Bit user will not know how coroutines work (or how Fluent Bit works). Something like this should intuitively make sense to people I think: "By default `concurrency` is set to `0` and records are flushed in Fluent Bit's single thread. This means that requests to Kinesis will block the execution of Fluent Bit as whole."

PettitWesley reviewed (changes requested):
Thanks for working on this- Comments from first read through, will go over again on the weekend

PettitWesley commented:
Did a second pass; couldn't find anything else to complain about.

zackwine commented:
I'll address the code review, and retest today

zackwine committed:
Address code review.  Update docs, and make retries configurable.

PettitWesley commented:
Ran the Kinesis integ tests, passed.

zackwine commented:
@PettitWesley  I ran several load tests with these changes, and verified `concurrency` and `concurrency_retires` work as expected.

I'm still seeing a limit of about 3000-3400 log lines per second before fluentbit starts falling behind, but overall it behaves much better under load (drops fewer records).

PettitWesley reviewed (approved):
@zackwine Please raise a PR on the main repo for this new feature 🥳

PettitWesley commented:
Also, once we get around to re-implementing this plugin in C, I'm going to ask you to run your performance tests on that. Hopefully it can hit higher throughput.

zackwine committed:
Merge master

zackwine committed:
Address code review items.

Added upper limit of 10 to concurrency.
Removed the unused records parameter from getPartitionKey.
Fixed improved error handling of configuration parameters.

Signed-off-by: Zack Wine <zwine@synamedia.com>

zackwine committed:
Address code review items (2).

Rename concurrency retires variable to concurrencyRetryLimit to be more consistent.
Update docs to mention concurrency option maximum value (10).

Signed-off-by: Zack Wine <zwine@synamedia.com>

zackwine committed:
Address code review.  Add data loss warning for concurrency.

zackwine committed:
Address code review.  Rename the concurrency feature experimental.

zackwine committed:
Clean up imports.

zackwine committed:
Fix unit tests.
